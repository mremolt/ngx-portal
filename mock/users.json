[
  {
    "id": 1,
    "email": "testuser1@example.com",
    "password": "secret"
  },
  {
    "id": 2,
    "email": "testuser2@example.com",
    "password": "secret"
  },

  {
    "id": 3,
    "email": "testuser3@example.com",
    "password": "secret"
  }
]
