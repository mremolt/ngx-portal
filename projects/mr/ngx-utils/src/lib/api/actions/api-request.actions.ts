import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

import { ApiActions, ApiRequestConfig } from '../types';

export enum ApiRequestActionTypes {
  ApiRequest = '[API] Request',
  ApiError = '[API] Error',
}

export class ApiRequest<T = any> implements Action {
  public readonly type = ApiRequestActionTypes.ApiRequest;

  constructor(public payload: ApiRequestConfig<T>) {}
}

export class ApiError implements Action {
  public readonly type = ApiRequestActionTypes.ApiError;

  constructor(public payload: HttpErrorResponse) {}
}

export class ApiGetAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('GET', url);
    super({ request, actions });
  }
}

export class ApiPostAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('POST', url, body);
    super({ request, actions });
  }
}

export class ApiPutAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('PUT', url, body);
    super({ request, actions });
  }
}

export class ApiDeleteAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('DELETE', url);
    super({ request, actions });
  }
}
