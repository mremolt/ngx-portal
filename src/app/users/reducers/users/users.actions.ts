import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { ApiErrorAction, ApiRequest, ApiStartAction, ApiSuccessAction } from '@mr/ngx-utils';

import { UserDto } from './user.dto';

export enum UsersActionTypes {
  LoadUsersStart = '[Users] Load Users Start',
  LoadUsersSuccess = '[Users] Load Users Success',
  LoadUsersError = '[Users] Load Users Error',
}

export class LoadUsersStart implements ApiStartAction {
  public readonly type = UsersActionTypes.LoadUsersStart;
  constructor(public payload?: any) {}
}

export class LoadUsersSuccess implements ApiSuccessAction<UserDto[]> {
  public readonly type = UsersActionTypes.LoadUsersSuccess;
  constructor(public payload: UserDto[]) {}
}

export class LoadUsersError implements ApiErrorAction {
  public readonly type = UsersActionTypes.LoadUsersError;
  constructor(public payload: HttpErrorResponse) {}
}

export class LoadUsers extends ApiRequest {
  constructor() {
    super({
      actions: { start: LoadUsersStart, success: LoadUsersSuccess, error: LoadUsersError },
      request: new HttpRequest('GET', 'users'),
    });
  }
}

export type UsersActions = LoadUsersStart | LoadUsersSuccess | LoadUsersError | LoadUsers;
