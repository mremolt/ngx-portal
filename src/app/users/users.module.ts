import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { StoreModule } from '@ngrx/store';
import * as fromUsers from './reducers/users/users.reducer';

@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    StoreModule.forFeature('users', fromUsers.reducer),
    // EffectsModule.forFeature([UsersEffects])
  ],
})
export class UsersModule {}
