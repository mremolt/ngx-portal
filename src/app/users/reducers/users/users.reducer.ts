import { createCollectionAdapter } from '@mr/ngx-utils';

import { UserDto } from './user.dto';
import { userSchema } from './user.schema';
import { UsersActionTypes, UsersActions } from './users.actions';

export const usersAdapter = createCollectionAdapter<UserDto>(userSchema);

export const initialState = usersAdapter.initialState;

export type State = Readonly<typeof initialState>;

export function reducer(state = initialState, action: UsersActions): State {
  switch (action.type) {
    case UsersActionTypes.LoadUsersStart:
      return { ...initialState, loading: true, loaded: false };

    case UsersActionTypes.LoadUsersSuccess:
      return { ...usersAdapter.reducers.upsertMany(state, action.payload, false), loaded: true };

    case UsersActionTypes.LoadUsersError:
      return { ...initialState, error: action.payload };

    default:
      return state;
  }
}
