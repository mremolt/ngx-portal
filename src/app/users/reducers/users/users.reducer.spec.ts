import { LoadUsersError, LoadUsersStart, LoadUsersSuccess } from './users.actions';
import { initialState, reducer } from './users.reducer';

describe('Users Reducer', () => {
  describe('unknown action', () => {
    it('should return the initial state', () => {
      const action = {} as any;
      const result = reducer(initialState, action);

      expect(result).toBe(initialState);
    });
  });

  describe('undefined state', () => {
    it('should return the initial state', () => {
      const action = {} as any;
      const result = reducer(undefined, action);

      expect(result).toBe(initialState);
    });
  });

  describe('LoadUsersStart', () => {
    it('matches a loading state', () => {
      expect(reducer(initialState, new LoadUsersStart())).toMatchSnapshot();
    });
  });

  describe('LoadUsersSuccess', () => {
    it('matches a success state', () => {
      expect(
        reducer(initialState, new LoadUsersSuccess([{ id: 12, name: 'test user' }] as any))
      ).toMatchSnapshot();
    });
  });

  describe('LoadUsersError', () => {
    it('matches a error state', () => {
      expect(
        reducer(initialState, new LoadUsersError({ message: 'Argh' } as any))
      ).toMatchSnapshot();
    });
  });
});
