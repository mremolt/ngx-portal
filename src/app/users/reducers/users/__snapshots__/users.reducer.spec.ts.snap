// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Reducer LoadUsersError matches a error state 1`] = `
Object {
  "dirty": false,
  "entities": Object {
    "collection": Object {},
  },
  "error": Object {
    "message": "Argh",
  },
  "loaded": false,
  "loading": false,
  "result": Array [],
  "selected": null,
  "updatedAt": null,
}
`;

exports[`Users Reducer LoadUsersStart matches a loading state 1`] = `
Object {
  "dirty": false,
  "entities": Object {
    "collection": Object {},
  },
  "error": null,
  "loaded": false,
  "loading": true,
  "result": Array [],
  "selected": null,
  "updatedAt": null,
}
`;

exports[`Users Reducer LoadUsersSuccess matches a success state 1`] = `
Object {
  "dirty": false,
  "entities": Object {
    "collection": Object {
      "12": Object {
        "id": 12,
        "name": "test user",
      },
    },
  },
  "error": null,
  "loaded": true,
  "loading": false,
  "result": Array [
    "12",
  ],
  "selected": null,
  "updatedAt": null,
}
`;
